dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+
+
+class MattesMIMetric(MattesBase):
+    def __init__(self, nbins=32, sampling_proportion=None):
+        r""" Initializes an instance of the Mattes MI metric
+
+        This class implements the methods required by Optimizer to drive the
+        registration process by making calls to the low level methods defined
+        in MattesBase.
+
+        Parameters
+        ----------
+        nbins : int, optional
+            the number of bins to be used for computing the intensity
+            histograms. The default is 32.
+        sampling_proportion : int in (0,100], optional
+            the percentage of voxels to be used for estimating the (joint and
+            marginal) intensity histograms. If None, dense sampling is used
+            (see notes regarding the difference between sparse and dense
+            sampling). The default is None.
+
+        Notes
+        -----
+        Since we use linear interpolation, images are not, in general,
+        differentiable at exact voxel coordinates, but they are differentiable
+        between voxel coordinates. When using sparse sampling, selected voxels
+        are slightly moved by adding a small random displacement within one
+        voxel to prevent sampling points to be located exactly at voxel
+        coordinates. When using dense sampling, this random displacement is
+        not applied.
+        """
+        super(MattesMIMetric, self).__init__(nbins)
+        self.sampling_proportion = sampling_proportion
+
+    def setup(self, transform, static, moving, static_grid2space=None,
+              moving_grid2space=None, prealign=None):
+        r""" Prepares the metric to compute intensity densities and gradients
+
+        The histograms will be setup to compute probability densities of
+        intensities within the minimum and maximum values of `static` and
+        `moving`
+
+        Parameters
+        ----------
+        transform: instance of Transform
+            the transformation with respect to whose parameters the gradient
+            must be computed
+        static : array, shape (S, R, C) or (R, C)
+            static image
+        moving : array, shape (S', R', C') or (R', C')
+            moving image
+        static_grid2space : array (dim+1, dim+1), optional
+            the grid-to-space transform of the static image. The default is
+            None, implying the transform is the identity.
+        moving_grid2space : array (dim+1, dim+1)
+            the grid-to-space transform of the moving image. The default is
+            None, implying the spacing along all axes is 1.
+        prealign : array, shape (dim+1, dim+1), optional
+            the pre-aligning matrix (an affine transform) that roughly aligns
+            the moving image towards the static image. If None, no
+            pre-alignment is performed. If a pre-alignment matrix is available,
+            it is recommended to directly provide the transform to the
```

Comment:
I think you are recommending using this `prealign` parameter rather than manual application of this affine before doing registration?  If so can you make this clear.

I agree with Eleftherios, better to avoid 'warping' for affine transformation in this docstring.

The name 'prealign' sounds like a flag rather than an affine.  Maybe 'starting_affine`?

I guess this affine maps from the output of `static_grid2space` to the input of `npl.inv(moving_grid2space)`?  That it is in world space?